### DISCUSSION TITLE
Discussion on RPC and API Collaborative Working Practices

### DISCUSSION THREAD

**Discussion by L2ncE** [cloudwego/hertz#590] [General]
See an issue on Lark, he said the following
>When using go for project development, I feel that some projects are pure HTTP or RPC, and some projects are RPC, and then an HTTP is proxied. The mode that everyone usually uses is more.

I think what he wants to say is, when developing a Go project, whether to use the HTTP directly, or use RPC + HTTP (BBF), or use RPC + Gateway (such as kong). I want to know what people think, or what scenarios they each apply to.

**Comment by li-jin-gou** [comment]
For different scenarios, there are two options:

- RPC + AGW
In the first scenario, our service needs to provide both RPC and API, e.g. openapi , we can use RPC and AGW, **such as kong or grpc-gateway.**
This allows us to add HTTP capabilities to our RPC service without having to do anything extra. It is suitable for open-api and less complex API scenarios.


- RPC + HTTP
When we use microservices, we split the service and we have to provide complex APIs for the client or the front, we need HTTP frameworks to assemble the data and all kinds of complex logic, this time I need to use HTTP frameworks such as **hertz**

If you are simply passing data through, I think using AGW is better than the HTTP framework,however, if you need to package data, I recommend using the HTTP framework (BackendsForFrontends).
